rfcs/enforce-legibility.md

```diff
@@ -0,0 +1,80 @@
+## Summary
+[summary]: #summary
+
+This RFC aims to introduce rules for limiting the usage of opens in OCaml code (`open Foo`) and automatically enforce these rules, using a dedicated tool.
+The infrastructure for enforcing these styles rules would also give way to the introduction of other legibility rules in the future if there is a consensus.
+For every such rule, the proposed infrastructure allows for an *incremental* process of compliance upgrade.
+
+## Motivation
+[motivation]: #motivation
+
+Currently, for a newcomer to either the whole code base or one of its components,
+it can be tedious to understand what is going on in the implementation of a feature. This can extend the development time of a new feature or facilitate the introduction of bugs if the code logic is not well understood.
+
+One of the reason is the heavy use of `open`s which impedes the reader's ability to understand where a value is coming from when used. Developer tools such as `merlin` can help locating the origin of a value but it quickly becomes tedious to review the origin of every ambiguous value, wheras directly qualifying the value instead of opening its module would immediatly resolve the problem.
+
+Of course, `open` is a useful tool and there are many cases where it makes sense to use it, such as for importing the infix operators of a module, or for library layering. This RFC aims to define simple, agreed on rules to determine when an `open` is problematic and either *remove* it (and re-qualify every imported value), *move* it closer to its usage, *transform* it into a local open or *restrain it* by explicitly expressing what symbols of the opened module are used in the source file.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+### Rules
+
+The rules are up to change, but we propose the following very basic rule as a first step, which only defines when an `open` should be completely removed.
+
+**Remove** an `open` if:
+- it is used 5 times or less in its scope
+
+**Except** if either:
+- the opened module is one of "Base", "Core" or "Core_kernel"
+- it is used to import operators
+- it is only used to import other modules (library layering)
+- its scope does not exceed 40 lines (roughly a screen)
+
+### Detection and enforcement
+
+This RFC relies on a tool, [ocaml-close](https://github.com/tweag/ocaml-close) (being built for the sake of this RFC) that can analyze and enforce open rules. This tool would be added as a dependency and installed in the mina switch.
```

Comment:
This repository is private, maybe worth mentioning that

Reply:
The repo is now public.